File: app/src/main/java/dev/bee/kanjianki/update/AutoUpdateWorker.java
package dev.bee.kanjianki.update;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import dev.bee.kanjianki.data.LocalStore;

public final class AutoUpdateWorker extends Worker {
    static UpdateClientFactory updateClientFactory = GitHubUpdater::androidClient;

    public AutoUpdateWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    @NonNull
    @Override
    public Result doWork() {
        return runFromStore(getApplicationContext(), automaticUpdateCheckerFactory(AutoUpdateWorker::checkAutomaticUpdate));
    }

    static UpdateCheckerFactory automaticUpdateCheckerFactory(AutomaticUpdateRunner runner) {
        return context -> () -> runner.check(context);
    }

    static GitHubUpdater.UpdateResult checkAutomaticUpdate(Context context) {
        return new GitHubUpdater(context, updateClientFactory.create(context)).checkDownloadAndInstall(GitHubUpdater.UpdateSource.AUTOMATIC);
    }

    static Result runFromStore(Context context, UpdateCheckerFactory checkerFactory) {
        Context appContext = context.getApplicationContext();
        try (LocalStore store = new LocalStore(appContext)) {
            LocalStore.AutoUpdateStatus status = store.autoUpdateStatus();
            if (!status.enabled || status.hasPendingUpdate()) {
                return Result.success();
            }
            return runAutoUpdate(true, false, checkerFactory.create(appContext));
        }
    }

    static Result runAutoUpdate(boolean enabled, boolean hasPendingUpdate, UpdateChecker checker) {
        if (!enabled || hasPendingUpdate) {
            return Result.success();
        }
        GitHubUpdater.UpdateResult result = checker.check();
        return result.retryable ? Result.retry() : Result.success();
    }

    interface UpdateChecker {
        GitHubUpdater.UpdateResult check();
    }

    interface UpdateCheckerFactory {
        UpdateChecker create(Context context);
    }

    interface AutomaticUpdateRunner {
        GitHubUpdater.UpdateResult check(Context context);
    }

    interface UpdateClientFactory {
        GitHubUpdater.UpdateClient create(Context context);
    }
}

File: app/src/main/kotlin/dev/bee/kanjianki/update/AutoUpdateWorker.kt
package dev.bee.kanjianki.update

import android.content.Context
import androidx.work.ListenableWorker.Result
import androidx.work.Worker
import androidx.work.WorkerParameters
import dev.bee.kanjianki.AndroidContainerProvider
import dev.bee.kanjianki.requireKaniContainer
import dev.bee.kanjianki.data.LocalStore
import dev.bee.kanjianki.updatecore.AutoUpdateRunPolicy

class AutoUpdateWorker internal constructor(
    context: Context,
    workerParams: WorkerParameters,
    private val containerProvider: AndroidContainerProvider,
) : Worker(context, workerParams) {
    constructor(context: Context, workerParams: WorkerParameters) : this(
        context,
        workerParams,
        AndroidContainerProvider { context.requireKaniContainer() },
    )

    override fun doWork(): Result {
        return containerProvider.get().openLocalStore().use { store ->
            runFromStore(
                store,
                applicationContext,
                automaticUpdateCheckerFactory(::checkAutomaticUpdate),
            )
        }
    }

    fun interface UpdateChecker {
        fun check(): GitHubUpdater.UpdateResult
    }

    fun interface UpdateCheckerFactory {
        fun create(context: Context?): UpdateChecker
    }

    fun interface AutomaticUpdateRunner {
        fun check(context: Context?): GitHubUpdater.UpdateResult
    }

    fun interface UpdateClientFactory {
        fun create(context: Context): GitHubUpdater.UpdateClient
    }

    companion object {
        @JvmField
        var updateClientFactory: UpdateClientFactory =
            UpdateClientFactory { context -> GitHubUpdater.androidClient(context) }

        @JvmStatic
        fun automaticUpdateCheckerFactory(runner: AutomaticUpdateRunner): UpdateCheckerFactory {
            return UpdateCheckerFactory { context -> UpdateChecker { runner.check(context) } }
        }

        @JvmStatic
        fun checkAutomaticUpdate(context: Context?): GitHubUpdater.UpdateResult {
            val updateContext = context!!
            return GitHubUpdater(updateContext, updateClientFactory.create(updateContext))
                .checkDownloadAndInstall(GitHubUpdater.UpdateSource.AUTOMATIC)
        }

        @JvmStatic
        fun runFromStore(context: Context, checkerFactory: UpdateCheckerFactory): Result {
            val appContext = context.applicationContext
            appContext.requireKaniContainer().openLocalStore().use { store ->
                return runFromStore(store, appContext, checkerFactory)
            }
        }

        private fun runFromStore(
            store: LocalStore,
            appContext: Context,
            checkerFactory: UpdateCheckerFactory,
        ): Result {
            val status = store.autoUpdateStatus()
            if (!AutoUpdateRunPolicy.shouldRun(status.enabled, status.hasPendingUpdate())) {
                return Result.success()
            }
            return runAutoUpdate(true, false, checkerFactory.create(appContext))
        }

        @JvmStatic
        fun runAutoUpdate(enabled: Boolean, hasPendingUpdate: Boolean, checker: UpdateChecker): Result {
            if (!AutoUpdateRunPolicy.shouldRun(enabled, hasPendingUpdate)) {
                return Result.success()
            }
            val result = checker.check()
            return workerResult(AutoUpdateRunPolicy.workerOutcome(result.retryable))
        }

        @JvmStatic
        fun workerResult(outcome: AutoUpdateRunPolicy.WorkerOutcome): Result {
            return if (outcome == AutoUpdateRunPolicy.WorkerOutcome.RETRY) {
                Result.retry()
            } else {
                Result.success()
            }
        }
    }
}
